arbor/backends/gpu/matrix_fine.cu

```diff
@@ -100,30 +100,40 @@ void solve_matrix_fine(
     T* rhs,
     T* d,
     const T* u,
-    const level* levels,
-    const unsigned* levels_start,
-    unsigned* num_matrix, // number of packed matrices = number of cells
-    unsigned* padded_size)
+    const level_metadata* levels_meta,
+    const fvm_index_type* levels_lengths,
+    const fvm_index_type* levels_parents,
+    const fvm_index_type* block_index,
+    fvm_index_type* num_matrix, // number of packed matrices = number of cells
+    fvm_index_type* padded_size)
 {
     const auto tid = threadIdx.x;
     const auto bid = blockIdx.x;
 
-    const auto first_level = levels_start[bid];
-    const auto num_levels  = levels_start[bid + 1] - first_level;
-    const auto block_levels = &levels[first_level];
+    const auto first_level = block_index[bid];
+    const auto num_levels  = block_index[bid + 1] - first_level;
+
+    const auto block_levels_meta = &levels_meta[first_level];
 
     // backward substitution
 
     for (unsigned l=0; l<num_levels-1; ++l) {
-        const auto& lvl = block_levels[l];
+        // Metadata for this level and the next level
+        const auto& lvl_meta = block_levels_meta[l];
+        const auto& next_lvl_meta = block_levels_meta[l+1];
+
+        // Addresses of the first elements of levels_lengths and levels_parents
+        // that belong to this level
+        const auto lvl_lengths = &levels_lengths[lvl_meta.level_data_index];
+        const auto lvl_parents = &levels_parents[lvl_meta.level_data_index];
```

Comment:
For me writing this using pointer arithmetic as `lvl_parents = levels_parents + lvl_meta.level_data_index;` makes more sense.

Reply:
Done.